src/porepy/numerics/fv/tpsa.py

```diff
@@ -0,0 +1,1379 @@
+        # discretization can be constructed. The actual discretization matrices are
+        # specified in Equation (A2.24) in the Tpsa paper (referring to the version
+        # found at https://arxiv.org/abs/2405.10390v2) for internal faces and (A2.28-31)
+        # for boundary faces. Compared to the description in the paper, the
+        # implementation uses a more explicit treatment of boundary conditions and
+        # prefers more explicit naming of mathematical operators, but should otherwise
+        # be equivalent.
+        #
+        # Specification of boundary conditions: Is the paper, the boundary conditions
+        # are given on the form
+        #
+        #       b(\sigma_k * n_k - g^u) = 2\mu (g^u - u_k).
+        #
+        # Here, b is considered a length scale and is also denoted \delta_k^j (see the
+        # bottom of page 10 in the arXiv version of the paper, link above). It is
+        # understood that b = 0 gives a Dirichlet condition, b = infinity gives a
+        # Neumann condition, and other values (positive and negative) form the Robin
+        # case. In PorePy, Dirichlet and Neumann conditions are set separately, while
+        # the Robin parameter, \alpha, is interpreted to be on the form
+        #
+        #       \sigma_k * n_k + \alpha u = g^u
+        #
+        # Thus, the PorePy \alpha corresponds to 2\mu / (\delta) in the paper.
+        #
+        # Implementation choices:
+        # 1. To improve oversight of the implementation, it was decided to move
+        #    computation of auxiliary quantities into static helper methods, and pass
+        #    data around through helper dataclasses (namedtuples or dictionaries could
+        #    also have been used). This is at times a bit awkward, but the alternative
+        #    was a more chaotic implementation.
+        # 2. Boundary conditions are implemented through a combination of filtering and
+        #    average maps. This is in contrast to the tpsa paper, which presents a
+        #    unified implementation of all types of boundary conditions based on the
+        #    average maps. While this approach likely could have been followed to a
+        #    larger degree, the more explicit treatment herein was preferred partly for
+        #    ease of implementation and interpretation, and (honestly) partly for
+        #    reasons of exhaustion on EK's part. For future reference, the most natural
+        #    alternative approach is to introduce one weight each for Dirichlet and
+        #    Neumann boundaries and let Robin be an average. The first step would be to
+        #    verify that this does not introduce division by zero somewhere.
+        #
+        # Known shortcomings and known unknowns:
+        # 1. The implementation of boundary conditions for the rotation variable covers
+        #    only Neumann and Dirichlet conditions. Robin conditions should be feasible,
```

Comment:
For future reference: it will most likely be rather straightforward now that a full implementation of boundary conditions is available, provided the right design choices are made for the implementation. It is not fully clear to me what that caveat  means in practice.